src/game/common/rts/energy.cpp

```diff
@@ -0,0 +1,191 @@
+/**
+ * @file
+ *
+ * @author tomsons26
+ *
+ * @brief
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "energy.h"
+#include "audiomanager.h"
+#include "gamelogic.h"
+#include "ini.h"
+#include "object.h"
+#include "player.h"
+#include "playerlist.h"
+
+Energy::Energy() : m_energyProduction(0), m_energyConsumption(0), m_frame(0), m_player(nullptr) {}
+
+void Energy::Init(Player *player)
+{
+    m_energyProduction = 0;
+    m_energyConsumption = 0;
+    m_player = player;
+    // TODO investigate, should this also clear m_frame?
+}
+
+int Energy::Get_Production() const
+{
+    if (g_theGameLogic->Get_Frame() < m_frame) {
+        return 0;
+    }
+    return m_energyProduction;
+}
+
+float Energy::Get_Energy_Supply_Ratio() const
+{
+    captainslog_dbgassert(m_energyProduction >= 0 && m_energyConsumption >= 0, "Negative Energy numbers!");
+
+    if (g_theGameLogic->Get_Frame() < m_frame) {
+        return 0.0f;
+    }
+
+    if (m_energyConsumption != 0) {
+        return (float)m_energyProduction / (float)m_energyConsumption;
+    }
+
+    return (float)m_energyProduction;
+}
+
+bool Energy::Has_Sufficient_Power() const
+{
+    if (g_theGameLogic->Get_Frame() < m_frame) {
+        return false;
+    }
+
+    return m_energyProduction >= m_energyConsumption;
+}
+
+void Energy::Adjust_Power(int amount, bool positive)
+{
+    if (amount != 0) {
```

Comment:
`amount > 0` inherintly is `amount != 0`, therefore this line is obsolete.

Reply:
Power adjustments occur with positives and negatives, assembly also has JNZ not JG so this is correct